### PR TITLE
[browser] Allow JS modules to be injected as pre-loaded assets

### LINKED ISSUE #89510: [browser] single file JavaScript and custom loader
### Description

I'm attempting to run `browser-wasm` with a minimal custom config under .NET 8:

```js
import { dotnet } from './dotnet.js'

const config = {
    "mainAssemblyName": "Test.dll",
    "assemblyRootFolder": "managed",
    "assets": [
	    {
	      "behavior": "dotnetwasm",
	      "name": "dotnet.native.wasm",
	    },
	    {
	      "behavior": "js-module-native",
	      "name": "dotnet.native.js",
	    },
	    {
	      "behavior": "js-module-dotnet",
	      "name": "dotnet.js",
	    },
	    {
	      "behavior": "js-module-runtime",
	      "name": "dotnet.runtime.js",
	    },
	    {
	      "behavior": "assembly",
	      "name": "System.Console.wasm",
	    },
	    {
	      "behavior": "assembly",
	      "name": "System.Linq.wasm",
	    },
	    {
	      "behavior": "assembly",
	      "name": "System.Private.CoreLib.wasm",
	    },
	    {
	      "behavior": "assembly",
	      "name": "Test.wasm",
	    }
    ]
};
const runtime = await dotnet.withConfig(config).create();
```
But getting nullref here: https://github.com/dotnet/runtime/blob/v8.0.0-preview.6.23329.7/src/mono/wasm/runtime/loader/assets.ts#L145C28

It worked in .NET 7 fine before.

### Reproduction Steps

Publish `browser-wasm` and try running it with a custom config under "AppBundle" dir.

### Expected behavior

No errors.

### Actual behavior

Error in mono_download_assets: TypeError: pendingDownloadInternal is null

### Regression?

Yes, it worked in .NET 7.

### Known Workarounds

None.

### Configuration

.NET v8.0.0-preview.6.23329.7
Windows 10
Local http server under AppBundle dir with `serve` under powershell.

### Other information

_No response_

### PR DESCRIPTION
**Pull request by pavelsavara**
- new `asset.moduleExports` field which could be pre-populated with ES6 exports of runtime modules
- demo in `browser-minimal-config\main.js`
- allow promise in `asset.buffer`

To address user feedback 
Fixes https://github.com/dotnet/runtime/issues/89510

### CODE REVIEW

**Comment by pavelsavara** [comment]
cc @Elringus

**Comment by Elringus** [comment]
This should work. Thank you!

File: src/mono/sample/wasm/browser-minimal-config/main.js [hunk 1]
```diff
@@ -1,24 +1,35 @@
...
     {
         name: "dotnet.native.js",
+        // demo dynamic import
+        moduleExports: import("./_framework/dotnet.native.js"),
         behavior: "js-module-native"
     },
-    {
-        name: "dotnet.js",
-        behavior: "js-module-dotnet"
-    },
```
Comment (ilonatommy) [inline src/mono/sample/wasm/browser-minimal-config/main.js:15]: Why removed?

File: src/mono/wasm/runtime/loader/assets.ts [hunk 2]
```diff
@@ -201,10 +201,11 @@ export async function mono_download_assets(): Promise<void> {
                 const asset = await downloadPromise;
                 if (asset.buffer) {
                     if (!skipInstantiateByAssetTypes[asset.behavior]) {
-                        mono_assert(asset.buffer && typeof asset.buffer === "object", "asset buffer must be array or buffer like");
+                        mono_assert(asset.buffer && typeof asset.buffer === "object", "asset buffer must be array or buffer like or promise of it");
```
Comment (ilonatommy) [inline src/mono/wasm/runtime/loader/assets.ts:204]:
```suggestion
                        mono_assert(asset.buffer && typeof asset.buffer === "object", "asset buffer must be array-like or buffer-like or promise of these");
```
I could not understand at first, it looks less confusing to me but I am not sure if it's correct.

**Review by ilonatommy — APPROVED** [review]

File: src/mono/sample/wasm/browser-minimal-config/main.js (hunk 1 above)
Comment (pavelsavara) [reply src/mono/sample/wasm/browser-minimal-config/main.js:15]: because it's not needed and not generated by build anymore either

**Pull request merged**